### ISSUE TITLE
'Bring my changes' discards all changes

### ISSUE THREAD

**Issue by Schaemelhout** [desktop/desktop#8579]
### Describe the bug

When switching branches I get the choice of bringing my uncommitted changes to the new branch or stashing them on the current branch.
When choosing to bring the changes to the new branch - **sometimes** - all my uncommitted changes are discarded...

### Version & OS

Version 2.2.2
Windows 10 Enterprise v10.0.17134 Build 17134

### Steps to reproduce the behavior

1. Switch to another branch while having uncommitted changes
2. Click on 'Bring my changes to branch-name'
3. Nothing happens, I am still on the previous branch, so I try again.
4. GitHub Desktop does not ask me to bring my changes to the new branch anymore, switched branches and discards my pending changes.
5. Changes pane shows '0 changed files'

### Logs
I have just examined the logs and it looks like an error occured behind the scenes where some changes would have been overridden by the checkout. This error was **not** shown in the GitHub Desktop UI.
```
2019-11-04T08:18:49.034Z - info: [ui] [Timing] Action 'checkout branch from list' for 'XXX' took 0.000s
2019-11-04T08:19:09.566Z - error: [ui] `git -c credential.helper= checkout --progress origin/XXX -b XXX --` exited with an unexpected code: 1.
stderr:
error: Your local changes to the following files would be overwritten by checkout:
	XXX/XXX.csproj
Please commit your changes or stash them before you switch branches.
Aborting

(The error was parsed as 37: Unable to switch branches as there are working directory changes which would be overwritten. Please commit or stash your changes.)
2019-11-04T08:19:10.067Z - info: [ui] [Timing] Action 'stash and checkout' for 'XXX' took 1.047s
2019-11-04T08:19:20.276Z - info: [ui] [Timing] Action 'checkout branch from list' for 'XXX' took 0.000s
2019-11-04T08:19:23.812Z - info: [ui] [Timing] Action 'stash and checkout' for 'XXX' took 1.219s
2019-11-04T08:19:34.713Z - info: [ui] [Timing] Action 'checkout branch from list' for 'XXX' took 0.847s
2019-11-04T08:19:38.138Z - info: [ui] [Timing] Action 'checkout branch from list' for 'XXX' took 0.931s
```

**Comment by steveward** [comment]
Thanks for the report @Schaemelhout. I wasn't able to reproduce this with a couple of quick test cases -- are there any extra details you may have about the state of your repository when this issue occurs? Are you seeing any merge conflicts?

There were changes made in https://github.com/desktop/desktop/issues/8085 to resolve a specific case when carrying over untracked files. Do you remember when you started experiencing this issue?

**Comment by Schaemelhout** [comment]
I have tried to reproduce this issue with a new GitHub repository without success.

1. Switch to branch "**dev**" from branch "**master**". Current branch is now "**dev**".
2. Push changes to branch "**master**" in a specific file (on another clone of the same repository or on another machine).
3. Make changes in the same file on branch "**dev**" to cause a conflict when trying to merge "**dev**" and "**master**".
4. Checkout branch "**master**" and bring those uncommitted changes.

I now get a conflict in my changes pane which should be **the correct behavior**. 

I remember experiencing this problem a couple of times in the last year or so (I don't think this feature of bringing changes to a new branch has been around for very long).

I have encountered this issue in the past on both Azure DevOps and Bitbucket repositories, but I can't seem to reproduce it on GitHub repositories. I will try to reproduce it on other platforms as well.

**Comment by outofambit** [comment]
@Schaemelhout are you still experiencing this with the older repository in your original post? if so, perhaps re-cloning it might resolve it? Let us know if you have any further problems!

**Comment by gitercn** [comment]
I have the same issue on my company's enterprise github.xxx.com in recent days. After click bring my changes, nothing happens and all my changes are lost. 

Oh I thought what a terrible day that hours of work are lost! But finally I figured the changes are stashed but not shown in the Github Desktop. I have to recovery them using command line:
`git stash list`
`git stash pop`

**Comment by outofambit** [comment]
@gitercn would you mind sharing your GitHub Desktop version and your system OS?

**Comment by clouddemo1** [comment]
@outofambit my GitHub Desktop version is 2.2.3, and my system OS is Windows 10 1903.

**Comment by gitercn** [comment]
@outofambit my Github Desktop version is 2.2.3 and my system OS is Windows 10.

**Comment by outofambit** [comment]
I've tried to repro this but have been unable to so far 😕. @gitercn could you share logs from this problem? did your changes happen to involve a `.csproj` file?

**Comment by Schaemelhout** [comment]
@outofambit It looks like I am unable to reproduce this issue even on my 'older' repository. 
I will let you know when it happens again in the future...

Thanks for your time anyway!

**Comment by Schaemelhout** [comment]
> I have the same issue on my company's enterprise github.xxx.com in recent days. After click bring my changes, nothing happens and all my changes are lost.
> 
> Oh I thought what a terrible day that hours of work are lost! But finally I figured the changes are stashed but not shown in the Github Desktop. I have to recovery them using command line:
> `git stash list`
> `git stash pop`

For reference: just experienced this issue again, and luckily the changes **do** seem to be stashed, so manually checking the stash is a great workaround.

**Comment by ChemicalXandco** [comment]
Just had this issue version is 2.2.4
Was able to fix by using `git stash pop`, thankfully

I think this issue should be reopened

**Comment by pavelevap** [comment]
I had the same issue with latest 2.2.4 version :-( Windows 7, 64 bit.

I am wondering if it is not related to origin branches? My issue was caused when I switched to `origin/somebranch-title`, but when I tried it by switching to local branch then it worked well.

**Comment by Peter92** [comment]
Also happened with 2.3.1 on Windows 10. Fortunately it wasn't a huge edit so I could rewrite the changes.

**Comment by immasood** [comment]
I faced the same issue a few days back.  and now I have lost a whole set of changes. any solution to recover those changes.

**Comment by OmarSherif96** [comment]
Hi All I have been able to find a way to solve this issue when I did bring my changes from branch `feature `to `master `by mistake before committing. 

On the master branch where the conflicts happened you will see something like this 
![image](https://user-images.githubusercontent.com/46057697/127996091-98778d66-4835-4902-84e7-6019d9620b81.png)

Accept all the incoming changes (Stashed changes) just to make sure nothing is lost and you have resolved your conflicts. Then in your terminal write `git reset --mixed` and then bring back your changes to the `feature `branch and again Accept all the incoming changes (Stashed changes). 

And that's it you are back to the branch with the changes done, hope that helps!

**Comment by FrancisPog1** [comment]
I also experience a similar problem. But in my case, this solves my problem.

The problem:
I'm in the main branch and I switched to the dummy-branch. I stashed my changes but if I go to the Github desktop, it didn't show all my stash and when I go back to the main branch in my VS Code, all changes that I made are gone. 

I did these to solve the problem:
1. I go to the folder of my project and I run the "Git Bash Here"
![image](https://github.com/desktop/desktop/assets/128720852/c99b0621-ca91-4e37-a130-1172f00ab71c)



2. I run this command to the bash git stash list and the result is this:
![image](https://github.com/desktop/desktop/assets/128720852/3a3ad30a-1b08-4ee1-9d71-bdba2c4d7d44)

3. After identifying the stash in the list, then I run this command "git stash pop stash@{1}"
![image](https://github.com/desktop/desktop/assets/128720852/f5e10e9c-aa90-4dca-b4c7-8b7ddf4d365d)

Note: If you all experience these errors using the terminal of VS code, then it would be better to use GIT BASH. In my case, this is the error:
![image](https://github.com/desktop/desktop/assets/128720852/47f3ed7c-c9c8-4616-8175-ae42f11f57de)
